### PR TITLE
[MM-34345] Convert switch channel provider

### PR DESCRIPTION
**Pull request by phi-h-nguyen**
Summary
This PR migrates 'components/suggestion/switch_channel_provider' and associated tests to TypeScript

Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/21156
JIRA: https://mattermost.atlassian.net/browse/MM-34345

### CODE REVIEW

**Comment by nevyangelova** [comment]
Hello @phi-h-nguyen can you please sign the CLA and then we can proceed with the process. Many thanks :)

File: components/suggestion/switch_channel_provider.tsx [hunk 1]
```diff
@@ -266,7 +303,9 @@ class SwitchChannelSuggestion extends Suggestion {
     }
 }
 
-function mapStateToPropsForSwitchChannelSuggestion(state, ownProps) {
+function mapStateToPropsForSwitchChannelSuggestion(state: GlobalState, ownProps: Readonly<any> & Readonly<{
+    children?: React.ReactNode;
```
Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.tsx:307]: you can add a type for ownProps

File: components/suggestion/switch_channel_provider.test.tsx [hunk 2]
```diff
@@ -142,33 +149,92 @@ describe('components/SwitchChannelProvider', () => {
         const users = [
             {
                 id: 'other_user',
-                display_name: 'other_user',
                 username: 'other_user',
+                create_at: 0,
```
Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.test.tsx:153]: i believe those changes are not needed in the scope for this PR

File: components/suggestion/switch_channel_provider.test.tsx [hunk 3]
```diff
@@ -142,33 +149,92 @@ describe('components/SwitchChannelProvider', () => {
...
+            scheme_id: '',
+            group_constrained: false,
         },
         {
             id: 'direct_other_user',
             type: 'D',
             name: 'current_user_id__other_user',
             display_name: 'other_user',
             delete_at: 0,
+            create_at: 0,
```
Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.test.tsx:218]: these too

File: components/suggestion/switch_channel_provider.tsx [hunk 4]
```diff
@@ -51,28 +55,61 @@ import {isGuest} from 'mattermost-redux/utils/user_utils';
...
 };
 
-const InsightsChannel = {
+const InsightsChannel: Channel = {
     id: 'insights',
     name: 'activity-and-insights',
     display_name: 'Insights',
     type: Constants.INSIGHTS,
     delete_at: 0,
+    create_at: 0,
```
Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.tsx:99]: can we make these props optional so that we don't have to add them to the tests or here. Alternatively you can pick only the props we are using from the Channel type 
https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys

File: components/suggestion/switch_channel_provider.tsx [hunk 5]
```diff
@@ -300,7 +339,19 @@ const ConnectedSwitchChannelSuggestion = connect(mapStateToPropsForSwitchChannel
 
 let prefix = '';
 
-function sortChannelsByRecencyAndTypeAndDisplayName(wrappedA, wrappedB) {
+type WrappedChannel = {
+    user?: UserProfile;
```
Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.tsx:343]: nit: can you move type declarations to the top of the file

Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.tsx:343]: nit: can you move type declarations to the top of the file

File: components/suggestion/switch_channel_provider.tsx [hunk 6]
```diff
@@ -499,7 +552,7 @@ export default class SwitchChannelProvider extends Provider {
         });
     }
 
-    userWrappedChannel(user, channel) {
+    userWrappedChannel(user: UserProfile, channel: Channel | null | undefined): WrappedChannel {
```
Comment (nevyangelova) [inline components/suggestion/switch_channel_provider.tsx:555]: Can specify undefined only here but AFAIK channel should always be present here?

File: packages/types/src/channels.ts [hunk 7]
```diff
@@ -33,7 +33,7 @@ export type Channel = {
     update_at: number;
     delete_at: number;
     team_id: string;
-    type: ChannelType;
+    type: ChannelType | string;
```
Comment (nevyangelova) [inline packages/types/src/channels.ts:36]: if the type you need is not currently present, please add it to the ChannelType interface

File: packages/types/tsconfig.json [hunk 8]
```diff
@@ -10,6 +10,7 @@
         "esModuleInterop": true,
         "allowSyntheticDefaultImports": true,
         "jsx": "react",
+        "types": ["node", "jest"],
```
Comment (nevyangelova) [inline packages/types/tsconfig.json:13]: why is this needed?

**Review by nevyangelova — CHANGES_REQUESTED** [review]
Thank you for your contribution, I have left some suggestions :) also you have committed two empty db files, can you please remove them? 

Perhaps the fork needs to be updated with our latest master too.

**Comment by M-ZubairAhmed** [comment]
Can you please fix the failing CI

**Comment by AshishDhama** [comment]
@phi-h-nguyen, we are looking forward to this. Could you please update us on the status?

**Comment by AshishDhama** [comment]
@phi-h-nguyen ping me once you are ready with changes

**Comment by lieut-data** [comment]
Heads up that as part of our efforts to move to a monorepo, we're closing out a number of older pull requests like this one to help streamline the effort.

If you'd like to preserve these changes -- even if you're not the original author! -- feel free to resubmit the pull request against the monorepo once it's ready. You can subscribe to [mattermost-server-issue-22420](https://www.github.com/mattermost/mattermost-server/issues/22420) for status updates on this effort.
